### ISSUE TITLE
Newer MBN ELF compatibility problems

### ISSUE THREAD

**Issue by copslock** [linneman/orct#2]
Hi,author of ORCT,thanks for developing such awesome tool to parse the Qualcomm MBN MCFG elf file,and there was some problem parsing the newer elf mbn used in android N platform devices
`
Error in transformation of 35-th mcfg-item occurred: mcfg item -> {:orct.mbn/ite
m_length 1010, :orct.mbn/item_type 4, :orct.mbn/attrib 25, :orct.mbn/sp_ops 0, :
orct.mbn/spare1 0} has wrong type 4!

[orct.mbn$extract_mcfg_items$fn__740 invoke mbn.clj 168]
[orct.mbn$extract_mcfg_items invoke mbn.clj 164]
[orct.mbn$parse_mbn_file_content invoke mbn.clj 200]
[orct.mbn$parse_mbn_data invoke mbn.clj 301]
[orct.core$print_file doInvoke core.clj 64]
[clojure.lang.RestFn invoke RestFn.java 494]
[orct.core$cli doInvoke core.clj 188]
[clojure.lang.RestFn applyTo RestFn.java 137]
[clojure.core$apply invoke core.clj 630]
[orct.core$_main doInvoke core.clj 228]
[clojure.lang.RestFn applyTo RestFn.java 137]
[orct.core main nil -1]`

Can you make some commit to support the newer format?
If you need some of the AMSS code,I can provide some of it,thanks a lot.
And is it possible to output the mbn's nv item file in item files mode? thanks

**Comment by copslock** [comment]
Thanks a lot!
I had tested the tool on MSM8992 MSM8936 MSM8996 MSM8953 MSM8940 platform's mcfg mbns,and the tool works seamlessly on 8953 and 8996 platform,and it was annoying about the wrong type format about the type mcfg-int-delete-nv-item & type mcfg-int-delete-efs-file nv items,which is declared not support by the current implementation on a few newer SDM450 platform&older msm8936 platform mbns,and some exception (details below) about the zlib-uncompress moudle when possible processing the compressed data in the earlier MSM8994 platform mbns,but works perfectly without uncompress the zilb compressed EFS items

`(def mcfg-int-nv-item                        0x01)
(def mcfg-int-efs-file                       0x02)
(def mcfg-int-sw-only                        0x03)
(def mcfg-int-delete-nv-item                 0x04)
(def mcfg-int-delete-efs-file                0x05)
(def mcfg-int-static-wo-efs-file             0x06)
(def mcfg-int-muxd-nv-item                   0x07)
(def mcfg-int-muxd-sw-only                   0x08)
(def mcfg-int-muxd-efs-file                  0x09)
(def mcfg-int-muxd-sw-only-efs-file          0x0a)
(def mcfg-int-data-profile                   0x0b)
(def mcfg-int-delete-data-profile            0x0c)
(def mcfg-int-static-wo-data-profile         0x0d)
(def mcfg-int-muxed-data-profile             0x0e)
(def mcfg-int-muxd-sw-only-data-profile      0x0f)
(def mcfg-int-config-item                    0x10)`

`No matching method found: alength
[clojure.lang.Reflector invokeMatchingMethod Reflector.java 80]
[clojure.lang.Reflector invokeStaticMethod Reflector.java 207]
[orct.utils$zlib_uncompress invoke utils.clj 275]
[orct.qcn_parser$parse_nv_efs_data$fn__546 invoke qcn_parser.clj 249]
[clojure.core.protocols$iter_reduce invoke protocols.clj 49]
[clojure.core.protocols$fn__6510 invoke protocols.clj 112]
[clojure.core.protocols$fn__6452$G__6447__6465 invoke protocols.clj 13]
[clojure.core$reduce invoke core.clj 6519]
[orct.qcn_parser$parse_nv_efs_data invoke qcn_parser.clj 240]
[orct.qcn_parser$subst_with_parsed_nv_efs_data invoke qcn_parser.clj 282]
[orct.qcn_printer$print_nv_item_set doInvoke qcn_printer.clj 204]
[clojure.lang.RestFn invoke RestFn.java 464]
[orct.core$print_file doInvoke core.clj 73]
[clojure.lang.RestFn invoke RestFn.java 494]
[orct.core$cli doInvoke core.clj 188]
[clojure.lang.RestFn applyTo RestFn.java 137]
[clojure.core$apply invoke core.clj 630]
[orct.core$_main doInvoke core.clj 228]
[clojure.lang.RestFn applyTo RestFn.java 137]
[orct.core main nil -1]`

There maybe more problems like the exceptions above because Qualcomm always update the NV/EFS items format or whatever makes the tool hard to parse the ELF with the same code method. And about the NVdefinitions, I just used the QXDM's parse format and manually make some definition as the format in some XML file. Besides,does the mcfg_build.pl helps anything？So far the SDM630&660&636 support is not very necessary.
[mcfg_sw_8994.tar.gz](https://github.com/linneman/orct/files/1905636/mcfg_sw_8994.tar.gz)
[mcfg_sw_SDM450_type_4_error.tar.gz](https://github.com/linneman/orct/files/1905646/mcfg_sw_SDM450_type_4_error.tar.gz)
[mcfg_sw_8936_type_5_error.tar.gz](https://github.com/linneman/orct/files/1905660/mcfg_sw_8936_type_5_error.tar.gz)
[NvDefinition.tar.gz](https://github.com/linneman/orct/files/1905662/NvDefinition.tar.gz)
[mcfg_sw_sdm630_version_not_surpport.tar.gz](https://github.com/linneman/orct/files/1905665/mcfg_sw_sdm630_version_not_surpport.tar.gz)

All in all ,thanks for developing such awesome tool to help manually active IMS-based Voice on LTE. It's hard for the community to support such proprietary files used in the Qualcomm platform.

Best regards,
QTRUN Quicktest Technology
www.qtrun.com

**Comment by copslock** [comment]
MBN Root/Carrier config XML&Group XML]
For MSM8998, SDM660, MDM9x50 platform used the root XML/Carrier config XML&Group XML
The root XML is:
modem_proc\mcfg\mcfg_gen\generic\China\CT\mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml
The group XMLs are:

modem_proc/mcfg/mcfg_gen/groups/mcfg/China/CT/mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml

modem_proc/mcs/configs/groups/China/CT/mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml

modem_proc/lte/configs/configs_xml/China/CT/mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml

modem_proc/mmcp/config/China/CT/mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml

modem_proc/uim/config/groups/China/CT/mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml

modem_proc/hdr/configs/China/CT/mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml

modem_proc/1x/configs/groups/China/CT/mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml

modem_proc/gps/gnss_tools/config/China/CT/mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml

modem_proc/wcdma/configs/groups/China/CT/mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml

modem_proc/datamodem/datacfg/xml/China/CT/mcfg_sw_gen_VoLTE_OpenMkt-Commercial.xml

modem_proc/ims/Configuration/MasterMBNConfiguration/China/CT/ims_master.xml

just some wiki about the newer format ：）

**Comment by linneman** [comment]
Hi Copslock,

Manu thanks for your support and the mbn files you have provided. I was able to fix the parsing operations now at least such that the tool does not throw an exception anymore. I hope the generated output is correct. As usual with free and open source source I cannot give any warranty so please treat the results with care.

I hope that this tool is nevertheless somehow useful for you.

Best Regards,
Otto

**Comment by copslock** [comment]
Well,looks like works perfectly on most MBNs right now.
But still the weird **alength** problem on MSM8994/MSM8992 with the Nvdeifinition below

Parsing result for file /root/miui_MI4c_5.12.4_b6d89f351d_5.
1/mbn/cmcc/mcfg_sw.mbn using schema definition Nv.xml

**No matching method found: alength**
[clojure.lang.Reflector invokeMatchingMethod Reflector.java 80]
[clojure.lang.Reflector invokeStaticMethod Reflector.java 207]
[orct.utils$zlib_uncompress invoke utils.clj 275]
[orct.qcn_parser$parse_nv_efs_data$fn__546 invoke qcn_parser.clj 249]
[clojure.core.protocols$iter_reduce invoke protocols.clj 49]
[clojure.core.protocols$fn__6510 invoke protocols.clj 112]
[clojure.core.protocols$fn__6452$G__6447__6465 invoke protocols.clj 13]
[clojure.core$reduce invoke core.clj 6519]
[orct.qcn_parser$parse_nv_efs_data invoke qcn_parser.clj 240]
[orct.qcn_parser$subst_with_parsed_nv_efs_data invoke qcn_parser.clj 282]
[orct.qcn_printer$print_nv_item_set doInvoke qcn_printer.clj 204]
[clojure.lang.RestFn invoke RestFn.java 464]
[orct.core$print_file doInvoke core.clj 73]
[clojure.lang.RestFn invoke RestFn.java 494]
[orct.core$cli doInvoke core.clj 188]
[clojure.lang.RestFn applyTo RestFn.java 137]
[clojure.core$apply invoke core.clj 630]
[orct.core$_main doInvoke core.clj 228]
[clojure.lang.RestFn applyTo RestFn.java 137]
[orct.core main nil -1]

I lookup the stackoverflow and someone explain as below

As the docstring says, alength works on Java™ arrays, such as a String[] or Integer[], which is definitely an incompatible type with Clojure lists or vectors, for which you want to use count:

user=> (def x '(1 2 3))
#'user/x
user=> (def xa (to-array x))
#'user/xa
user=> (class x)
clojure.lang.PersistentList
user=> (class xa)
[Ljava.lang.Object;
user=> (alength xa)
3
user=> (alength x) 
java.lang.IllegalArgumentException: No matching method found: alength (NO_SOURCE_FILE:0)
user=> (count x)
3
[Ljava.lang.Object; is the weird way toString is defined to output for native Object arrays.

In the code 

> `alength data`

the [data] seems like not defined before or something wrong when parsing the zlib-compressed data caused the problem.
[mcfg_sw_MSM8992.tar.gz](https://github.com/linneman/orct/files/1910170/mcfg_sw_MSM8992.tar.gz)
[NvDefinition_mod.tar.gz](https://github.com/linneman/orct/files/1910172/NvDefinition_mod.tar.gz)

If have any extra time,hope fix the last problem for working at least a while :)
Besides,honestly speaking,i am not hoping anyone could code some wonderful tool for MBN elf files,but here you did,That was really awesome.
Are you participated in any of the development of qualcomm devices?I can not image anyone else can have an interest in this area :)

Yours sincerely

**Comment by linneman** [comment]
yes, it seems like that the handling of compressed efs items within mbn files was completely broken.

Anyway, my consideration is that here is a wrong declaration in the schema definition you used where one of the keys obviously does not include compressed data. Instead of just trowing and exception the parsing process is now continued and an error message is generated. In the given case you get:

uncompression error 'incorrect header check' occurred in path: /nv/item_files/rfnv/00028447, key :23!

When you disable compression in the schema correspondingly this message disappears. In case a data chunk cannot be uncompressed it is used as it is. So in given case the output shall be the same.

Hope that helps and Best Regards,

Otto
